packages/dds/migration-shim/src/migrationShim.ts

```diff
@@ -2,25 +2,85 @@
  * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
  * Licensed under the MIT License.
  */
-
+import { TypedEventEmitter } from "@fluid-internal/client-utils";
+import {
+	type IEvent,
+	type IFluidHandle,
+	type IFluidLoadable,
+} from "@fluidframework/core-interfaces";
 import {
 	type IChannelAttributes,
 	type IChannel,
 	type IChannelServices,
+	type IFluidDataStoreRuntime,
+	type IChannelFactory,
 } from "@fluidframework/datastore-definitions";
 import {
 	type IExperimentalIncrementalSummaryContext,
 	type IGarbageCollectionData,
 	type ITelemetryContext,
 	type ISummaryTreeWithStats,
 } from "@fluidframework/runtime-definitions";
-import { type IFluidHandle, type IFluidLoadable } from "@fluidframework/core-interfaces";
+import { type SharedObject } from "@fluidframework/shared-object-base";
+
+/**
+ * Interface for migration events.
+ */
+export interface IMigrationEvent extends IEvent {
+	/**
+	 * Event that is emitted when the migration is complete.
+	 */
+	(event: "migrated", listener: () => void);
+}
+
+/**
+ * Interface for migration operation.
+ */
+export interface IMigrationOp {
+	/**
+	 * Type of the migration operation.
+	 */
+	type: "hotSwap";
+	/**
+	 * Old channel attributes.
+	 */
+	oldAttributes: IChannelAttributes;
+	/**
+	 * New channel attributes.
+	 */
+	newAttributes: IChannelAttributes;
+}
 
 /**
  * Create skeleton Migration Shim that can hot swap from one DDS to a new DDS.
  */
-export class MigrationShim implements IChannel {
-	public id!: string;
+export class MigrationShim<TOld extends SharedObject, TNew extends SharedObject>
+	extends TypedEventEmitter<IMigrationEvent>
+	implements IChannel
+{
+	public constructor(
+		public readonly id: string,
+		private readonly runtime: IFluidDataStoreRuntime,
+		private readonly oldFactory: IChannelFactory, // Should this be a legacy shared tree factory only?
+		private readonly newFactory: IChannelFactory, // Should this be a new shared tree factory only?
+		private readonly populateNewSharedObjectFn: (
+			oldSharedObject: TOld,
+			newSharedObject: TNew,
+		) => void,
```

Comment:
I think it would make sense to go ahead and make these legacy/new shared tree specific already - will probably cut down on effort re: the type params.